doc/source/dev/api-dev/nan_policy.rst

```diff
@@ -0,0 +1,168 @@
+collection of 1-d arrays, with the ``axis`` parameter specifying the axis
+along which the reduction is to be applied.  If, say::
+
+    func([1, 3, 4])     -> 10.0
+    func([2, -3, 8, 2]) ->  4.2
+    func([7, 8])        ->  9.5
+    func([])            -> -inf
+
+then::
+
+    func([[  1, nan,   3,   4],
+          [  2,  -3,   8,   2],
+          [nan,   7, nan,   8],
+          [nan, nan, nan, nan]], nan_policy='omit', axis=-1)
+
+must give the result::
+
+    np.array([10.0, 4.2, 9.5, -inf])
+
+
+Edge cases
+----------
+A function that implements the ``nan_policy`` parameter should gracefully
+handle the case where *all* the values in the input array(s) are ``nan``.
+The basic principle described above still applies::
+
+    func([nan, nan, nan], nan_policy='omit')
+
+should behave the same as::
+
+    func([])
+
+In practice, when adding ``nan_policy`` to an existing function, it is
+not unusual to find that the function doesn't already handle this case
+in a well-defined manner, and some thought and design may have to be
+applied to ensure that it works.  The correct behavior (whether that be
+to return ``nan``, return some other value, raise an exception, or something
+else) will be determined case-by-case.
+
+
+Why doesn't ``nan_policy`` also apply to ``inf``?
+--------------------------------------------------
+Although we learn in grade school that "infinity is not a number", the
+floating point values ``nan`` and ``inf`` are qualitatively different.
+The values ``inf`` and ``-inf`` act much more like regular floating
+point values than ``nan``.
+
+* One can compare ``inf`` to other floating point values and it behaves
+  as expected, e.g. ``3 < inf`` is True.
+* For the most part, arithmetic works "as expected" with ``inf``,
+  e.g. ``inf + inf = inf``, ``-2*inf = -inf``, ``1/inf = 0``,
+  etc.
+* Many existing functions work "as expected" with ``inf``:
+  ``np.log(inf) = inf``, ``np.exp(-inf) = 0``,
+  ``np.array([1.0, -1.0, np.inf]).min() = -1.0``, etc.
+
+So while ``nan`` almost always means "something went wrong" or "something
+is missing", ``inf`` can in many cases be treated as a useful floating
+point value.
+
+It is also consistent with the NumPy ``nan`` functions to not ignore
+``inf``::
+
+    >>> np.nanmax([1, 2, 3, np.inf, np.nan])
+    inf
+    >>> np.nansum([1, 2, 3, np.inf, np.nan])
+    inf
+    >>> np.nanmean([8, -np.inf, 9, 1, np.nan])
+    -inf
+
+
+Subtleties of ``nan_policy='propagate'``
+----------------------------------------
+In many case, ``nan_policy='propagate'`` is trivial to implement: just
+execute the code in the function without checking for ``nan``, and let the
+chips fall where they may.  There are two potential problems with this:
+
+* It may result in warnings that, for the user, are unexpected. E.g. "Why
+  are you warning me about invalid values in double scalars when I have
+  already told you to propagate the ``nan`` values?"
```

Comment:
You're right, this should provide clearer guidance.  Or maybe no guidance--perhaps my attempt to discuss the subtleties in this document is premature and can be removed.

The motivating example--indeed, the only example at the moment--is the issue reported in https://github.com/scipy/scipy/issues/7818.  The uniform filter is just a sliding windowed mean of the data.  There are at least three ways such a calculation could be implemented, and each one propagates `nan` differently.  In the "naive" implementation, the `nan`s in the output are limited to a short section with length equal to the window length.  For the implementation in `ndimage.uniform_filter`, the `nan` in the output propagates from the first occurrence of the input `nan` to the end of the array.  If the filter was calculated using the FFT, all the output values would be `nan`.  Mathematically (and in the absence of `nan`), all three implementations are equivalent--and a user shouldn't have to care about how that calculation is actually implemented.  But when we consider propagating a `nan`, the implementation details can become part of the API.  For the case of the uniform filter, do we want to promise that `nan` will propagate according to the naive implementation (as the user in that issue expected)?  This might mean checking for `nan` in advance, or checking for it during the computation, and switching to the naive implementation if a `nan` is encountered.  Or do we not bother trying to promise exactly how `nan`s will propagate, and instead state that the behavior should not be relied on because it is implementation-dependent and might change if the underlying implementation changes?  I'm not sure at the moment, which is why I think maybe I should take this out of the document.

Regarding warnings:  I've been trying to cook up an example or two where I can say "Here's a naive implementation of the 'propagate' behavior for a function that generates an undesirable warning", but I haven't had much luck.  All my simple examples fail to generate a warning, and I don't recall if I originally had an example that motivated me to write this.  So yet another reason to perhaps just take this section out.

Reply:
> You're right, this should provide clearer guidance. Or maybe no guidance--perhaps my attempt to discuss the subtleties in this document is premature and can be removed.

 > The motivating example
> ... [eliding full quoting of a long section]
> I'm not sure at the moment, which is why I think maybe I should take this out of the document.

I'm in agreement on taking this part out of the document, if examples come up later we can always add this language, or some updated version of it, into the document.
 
> Regarding warnings: I've been trying to cook up an example or two where I can say "Here's a naive implementation of the 'propagate' behavior for a function that generates an undesirable warning", but I haven't had much luck. All my simple examples fail to generate a warning, and I don't recall if I originally had an example that motivated me to write this. So yet another reason to perhaps just take this section out.

Hmm, I can't really think of any obvious example either so maybe is best to just take it out for now.